### PR TITLE
HDDS-1530. Freon support big files larger than 2GB and add --bufferSize and --validateWrites options.

### PR DESCRIPTION
**Pull request by iamcaoxudong**
https://issues.apache.org/jira/browse/HDDS-1530
Ozone: Freon: Support big files larger than 2GB and add "--bufferSize" and "--validateWrites" options.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 41 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 427 | trunk passed |
| +1 | compile | 196 | trunk passed |
| +1 | checkstyle | 52 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 890 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 145 | trunk passed |
| 0 | spotbugs | 271 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 478 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 407 | the patch passed |
| +1 | compile | 204 | the patch passed |
| +1 | javac | 204 | the patch passed |
| +1 | checkstyle | 55 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 696 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 122 | the patch passed |
| +1 | findbugs | 441 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 165 | hadoop-hdds in the patch failed. |
| -1 | unit | 1189 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 37 | The patch does not generate ASF License warnings. |
| | | 5673 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.TestContainerReplication |
|   | hadoop.ozone.scm.node.TestQueryNode |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-830/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/830 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 1c3367a285ee 3.13.0-153-generic #203-Ubuntu SMP Thu Jun 14 08:52:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / c183bd8 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/1/testReport/ |
| Max. process+thread count | 5268 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/tools U: hadoop-ozone/tools |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java [hunk 1]
```diff
@@ -228,8 +243,20 @@ public Void call() throws Exception {
...
+
+    // Compute the common initial digest for all keys without their UUID
+    if (validateWrites) {
+      commonInitialMD = DigestUtils.getDigest(DIGEST_ALGORITHM);
+      int uuidLength = UUID.randomUUID().toString().length();
+      keySize = Math.max(uuidLength, keySize);
+      for (long nrRemaining = keySize - uuidLength; nrRemaining > 0;
+          nrRemaining -= bufferSize) {
+        int curSize = (int)Math.min(bufferSize, nrRemaining);
+        commonInitialMD.update(keyValueBuffer, 0, curSize);
```
Comment (xiaoyuyao) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: The keyValueBuffer is only generated randomly once before the loop. Repeatedly calculate the same content won't change the digest here. Some of the commonInitalMD.update() cycles can be saved here.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java [hunk 2]
```diff
@@ -622,7 +642,11 @@ public void run() {
                 try (Scope writeScope = GlobalTracer.get()
                     .buildSpan("writeKeyData")
                     .startActive(true)) {
-                  os.write(keyValue);
+                  for (long nrRemaining = keySize - randomValue.length;
+                        nrRemaining > 0; nrRemaining -= bufferSize) {
+                    int curSize = (int)Math.min(bufferSize, nrRemaining);
+                    os.write(keyValueBuffer, 0, curSize);
```
Comment (xiaoyuyao) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Can we generate random string here and update the digest inline up to the buffersize? 
This way, we will not be writing the same buffer repeatedly that will be cached in OS buffer/cache.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java [hunk 3]
```diff
@@ -228,8 +243,20 @@ public Void call() throws Exception {
       init(freon.createOzoneConfiguration());
     }
 
-    keyValue =
-        DFSUtil.string2Bytes(RandomStringUtils.randomAscii(keySize - 36));
+    keyValueBuffer = DFSUtil.string2Bytes(
+        RandomStringUtils.randomAscii(bufferSize));
```
Comment (arp7) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Let's just write zeroes by default. Random data generation may itself become a bottleneck. We could add. a separate option for random data later.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java [hunk 4]
```diff
@@ -136,7 +141,20 @@
       description = "Specifies the size of Key in bytes to be created",
       defaultValue = "10240"
   )
-  private int keySize = 10240;
+  private long keySize = 10240;
+
+  @Option(
+      names = "--validateWrites",
+      description = "Specifies whether to validate keys after writtting"
```
Comment (arp7) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: typo: writtting -> writing.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java [hunk 5]
```diff
@@ -136,7 +141,20 @@
...
+  )
+  private boolean validateWrites = false;
+
+  @Option(
+      names = "--bufferSize",
+      description = "Specifies the buffer size while writting",
```
Comment (arp7) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: typo: writting -> writing.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java [hunk 6]
```diff
@@ -254,7 +282,7 @@ public Void call() throws Exception {
       writeValidationFailureCount = 0L;
 
       validationQueue =
-          new ArrayBlockingQueue<>(numOfThreads);
+          new LinkedBlockingQueue<>();
```
Comment (arp7) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: nitpick: don't need a line break here.

Comment (iamcaoxudong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Thanks， I'll correct it.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java (hunk 5 above)
Comment (iamcaoxudong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Thanks， I'll correct it.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java (hunk 4 above)
Comment (iamcaoxudong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Thanks， I'll correct it.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java (hunk 2 above)
Comment (iamcaoxudong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Thanks， I'll correct it,  after that every key will have an unique UUID.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java (hunk 1 above)
Comment (iamcaoxudong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Thank you, you are right, this is a place that can be optimized, I will correct it.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java (hunk 3 above)
Comment (iamcaoxudong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: Thank you, but the generation of random data may not take too much time, just like xiaoyuyao said below, repeatedly calculate the same content won't change the digest, so as an optimization, MessageDigest.update() is only required for the first calculation and the last calculation. this is quickly indeed.

**Comment by iamcaoxudong** [comment]
@arp7 @xiaoyuyao

**Comment by iamcaoxudong** [comment]
@arp7 @xiaoyuyao 
Thank you for your reviewing, I seriously considered your comments and conducted some related tests. I replied your reviewing opinions above.

I have just pushed a new commit and corrected some bugs. Please review it again, thank you!

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 76 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 890 | trunk passed |
| +1 | compile | 391 | trunk passed |
| +1 | checkstyle | 108 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 1059 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 158 | trunk passed |
| 0 | spotbugs | 305 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 501 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 498 | the patch passed |
| +1 | compile | 287 | the patch passed |
| +1 | javac | 287 | the patch passed |
| +1 | checkstyle | 86 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 713 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 156 | the patch passed |
| +1 | findbugs | 508 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 183 | hadoop-hdds in the patch failed. |
| -1 | unit | 1249 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 56 | The patch does not generate ASF License warnings. |
| | | 7115 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.common.impl.TestContainerPersistence |
|   | hadoop.ozone.client.rpc.TestBCSID |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-830/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/830 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 285dca4b5082 3.13.0-153-generic #203-Ubuntu SMP Thu Jun 14 08:52:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a771e2a |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/2/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/2/testReport/ |
| Max. process+thread count | 5141 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/tools U: hadoop-ozone/tools |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-830/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java (hunk 1 above)
Comment (xiaoyuyao) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: The explaination makes sense to me. Thanks.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java (hunk 2 above)
Comment (xiaoyuyao) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/RandomKeyGenerator.java]: You are right. No issue at the socket layer. I'm thinking of the DN side, the chunk files of the same key being written could be the same in this scheme. That might increase the write performance compared with 2GB fully random chunks.  As long we use it consistently, it should be fine. Later on, we will can an option to write 0 only by default and  random up to buffersize when a parameter is specified.

**Comment by xiaoyuyao** [comment]
+1, I will merge/commit this shortly.

**Pull request merged**